modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,265 @@
+    crop_ = crop;
+}
+
+void Model::setInputSwapRB(bool swapRB) {
+    swapRB_ = swapRB;
+}
+
+std::pair<int, float> Model::classify(const Mat& frame) {
+    Mat blob = blobFromImage(frame, scale_, Size(width_, height_), mean_, swapRB_, crop_);
+    setInput(blob);
+    Mat out = forward();
+    std::vector<float> result = out.reshape(1, 1);
+    std::vector<float>::iterator iter = std::max_element(result.begin(), result.end());
+
+    std::pair<int, float> prediction(iter - result.begin(), *iter);
+    return prediction;
+}
+
+void Model::detect(Mat& frame, std::vector<int>& classIds,
+                   std::vector<float>& confidences, std::vector<Rect2d>& boxes,
+                   float confThreshold, float nmsThreshold)
+{
+    Mat blob = blobFromImage(frame, scale_, Size(width_, height_), mean_, swapRB_, crop_);
+    setInput(blob);
+    if (getLayer(0)->outputNameToIndex("im_info") != -1) {  // Faster-RCNN or R-FCN
+        resize(frame, frame, Size(width_, height_));
+        Mat imInfo = (Mat_<float>(1, 3) << frame.rows, frame.cols, 1.6f);
+        setInput(imInfo, "im_info");
+    }
+
+    std::vector<Mat> detections;
+    forward(detections);
+    CV_Assert(detections.size() > 0);
+
+    std::vector<String> layerNames = getLayerNames();
+    int lastLayerId = getLayerId(layerNames.back());
+    Ptr<Layer> lastLayer = getLayer(lastLayerId);
+
+    std::vector<int> predClassIds;
+    std::vector<Rect2d> predBoxes;
+    std::vector<float> predConf;
+    if (lastLayer->type == "DetectionOutput") {
+        for (int i = 0; i < detections.size(); ++i)
+        {
+            for (int j = 0; j < detections[i].size[2]; j++)
+            {
+                std::vector<Range> coord = { Range::all(), Range::all(), Range(j, j + 1), Range::all() };
+                Mat detect;
+                detections[i](coord).copyTo(detect);
+                std::vector<float> detection = detect.reshape(1, 1);
+                float conf = detection[2];
+                if (conf > confThreshold)
+                {
+                    float left = detection[3];
+                    float top = detection[4];
+                    float right = detection[5];
+                    float bottom = detection[6];
+                    float width = right - left + 1;
+                    float height = bottom - top + 1;
+
+                    classIds.push_back(static_cast<int>(detection[1]));
+                    boxes.push_back(Rect2d(left, top, width, height));
+                    confidences.push_back(conf);
+                }
+            }
+        }
+    } else if (lastLayer->type == "Region") {
+        for (int i = 0; i < detections.size(); ++i) {
+            for (int j = 0; j < detections[i].rows; ++j) {
+
+                Mat slice = detections[i].row(j).colRange(5, detections[i].cols);
+                std::vector<float> scores = slice.reshape(1, 1);
+                std::vector<float>::iterator classId = std::max_element(scores.begin(), scores.end());
```

Comment:
Please use `cv::minMaxLoc`. Otherwise there is extra memory copy at `std::vector<float> scores = slice.reshape(1, 1);`.